src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            // If we're in draining mode, short circut here.
+            if let Some(to_drain) = &mut self.draining {
+                if let Some((line, src)) = to_drain.pop() {
+                    return Ok(Async::Ready(Some((line, src))));
+                } else {
+                    return Ok(Async::Ready(None));
+                }
+            }
+
+            // Check for keys that have hit their timeout.
+            while let Ok(Async::Ready(Some(expired_key))) = self.timeouts.poll() {
+                self.expired.push_back(expired_key.into_inner());
+            }
+
+            match self.inner.poll() {
+                Ok(Async::Ready(Some((line, src)))) => {
+                    // Handle the incoming line we got from `inner`. If the
+                    // handler gave us something - return it, otherwise continue
+                    // with the flow.
+                    if let Some(val) = self.handle_line(line, src) {
+                        return Ok(Async::Ready(Some(val)));
+                    }
+                }
+                Ok(Async::Ready(None)) => {
+                    // We got `None`, this means the `inner` stream has ended.
+                    // Start flushing all existing data, stop polling `inner`.
+                    self.draining =
+                        Some(self.buffers.drain().map(|(k, v)| (v.into(), k)).collect());
+                }
+                Ok(Async::NotReady) => {
+                    // We didn't get any lines from `inner`, so we just give
+                    // a line from the expired lines queue.
+                    if let Some(key) = self.expired.pop_front() {
+                        if let Some(buffered) = self.buffers.remove(&key) {
+                            return Ok(Async::Ready(Some((buffered.freeze(), key))));
+                        }
+                    }
+
+                    return Ok(Async::NotReady);
+                }
+                Err(()) => return Err(()),
+            };
+        }
+    }
+}
+
+impl<T> LineAgg<T>
+where
+    T: Stream<Item = (Bytes, Filename), Error = ()>,
+{
+    /// Handle line, if we have something to output - return it.
+    fn handle_line(&mut self, line: Bytes, src: Filename) -> Option<(Bytes, Filename)> {
+        // Check if we already have the buffered data for the source.
+        if self.buffers.contains_key(&src) {
+            let condition_matched = self.config.condition_pattern.is_match(line.as_ref());
+            match self.config.mode {
+                // All consecutive lines matching this pattern are included in
+                // the group.
+                Mode::ContinueThrough => {
+                    if condition_matched {
+                        let buffered = self
+                            .buffers
+                            .get_mut(&src)
+                            .expect("already asserted key is present");
```

Comment:
Can this pattern be replaced by using the `HashMap` `Entry` API above to avoid the double `get` plus `expect`?

Reply:
See https://github.com/timberio/vector/pull/1852#discussion_r381934319